# APIs and Python

## The Many Use Cases for APIs
APIs can actually be used for many things - not just for retrieving information. Twilio has an API that allows you to write a script to call their API to send text messages to people. GitHub has an API for creating new repositories. Many services have an API allowing a computer to automate a task that a person might otherwise have to do through a website - whether uploading a photo to Flickr, searching for a company name in a state database or getting a list of garbage collection times for a municipality.

### _Learning Goals:_<br/>
  - Identify and discuss APIs
  - Discuss and explain different request (GET, POST, PUT, DELETE) and CRUD operations
  - Explore the attributes of a response object
  - Check the status of a request and interpret status codes
  - Access data from an API using the requests library
  - Create a pandas dataframe from the data returned from an API and visualize the data
  - If time, target an API of their own!!!

## Limitations of APIs
When working with APIs, there are some limitations you have to be aware of - especially relating to scope and scale.

* **Scope** - Just because a company has an API and has information, it doesn't mean you can get all of the information through their API. 

* **Scale** - Some APIs are provided for free as a public service. Others you have to pay for, or allow you to perform activities (like sending a text message) that you pay for. Make sure that you know what the rate limits are and that your use case isn't going to need more API calls than you will be able to make.

Let's take a look at the `requests` library and it's uses.

### "Requests is the only Non-GMO HTTP library for Python, safe for human consumption."

![img-req](requests-pic.jpeg)

Straight from the `requests` [documentation](https://pypi.org/project/requests/)

![nongmao](requests_med.png)

"Requests allows you to send organic, grass-fed HTTP/1.1 requests, without the need for manual labor. "<br>
[More requests documentation](https://2.python-requests.org//en/v0.10.6/api/)

### Let's get started!

In [1]:
import requests

In [ ]:
#!pip install requests

### Types of requests

We will mostly use GET, but there are other options

Let's say you are looking at an API for a car rental company like Hertz, or maybe even Zipcar.

The following different requests will generate these different responses.

|Request        | Result        |
|-------------|-------------|
|GET /stores/   | User sees the list of stores|
|GET /rentals/  | User sees the history of car rentals|
|POST /rentals/ | User rents a car|
|PUT /rentals/{id}/| User changes destination store|
|DELETE /rentals/{id}/| User cancels the active car rental|




In [2]:
rental_json = [{
    "id": 1,
    "name": "John St",
    "location": [29, 95],
    "available_cars_quantity": 10
},
{
    "id": 2,
    "name": "Scott St.",
    "location": [29.5, 95.7],
    "available_cars_quantity": 2
}]

In [9]:
#create a post request from https://httpbin.org/post and view it in json format

r = requests.post('https://httpbin.org/post', data=rental_json[1]) #put url and data = {'key':'value'} format here in the bananas!
r.json()

{'args': {},
 'data': '',
 'files': {},
 'form': {'available_cars_quantity': '2',
  'id': '2',
  'location': ['29.5', '95.7'],
  'name': 'Scott St.'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Content-Length': '73',
  'Content-Type': 'application/x-www-form-urlencoded',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.19.1'},
 'json': None,
 'origin': '65.207.79.74, 65.207.79.74',
 'url': 'https://httpbin.org/post'}

### Request Class and Attributes

In [10]:
#create a GET request from https://api.github.com/events and view it in json format

r = requests.get('https://api.github.com/events') #put url here in the bananas!
type(r)

requests.models.Response

Let's check the `attributes` of a `Response` object. <br>
https://2.python-requests.org//en/v0.10.6/api/

Some key attributes:

- url
- text
- status_code
- headers
- error

In [13]:
r.headers

{'Server': 'GitHub.com', 'Date': 'Wed, 15 May 2019 18:47:15 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Status': '403 Forbidden', 'X-RateLimit-Limit': '60', 'X-RateLimit-Remaining': '0', 'X-RateLimit-Reset': '1557946825', 'X-GitHub-Media-Type': 'github.v3; format=json', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '1; mode=block', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'Content-Security-Policy': "default-src 'none'", 'Content-Encoding': 'gzip', 'X-GitHub-Request-Id': '8F80:3970:69C115:D49E21:5CDC5EB3'}

### Checking out the status of your request

In [17]:
#get the status of your post request
r.status_code

403

#### Types of status codes
1xx - Informational responses

2xx - Success
- 200 OK
- 201 Created
- 204 No Content

3xx - Redirection

- 301 Moved Permanently (permanent URL redirection)
- 304 Not Modified (A conditional GET or HEAD request has been received and would have resulted in a 200 OK response if it were not for the fact that the condition evaluated to false.)

4xx - Client errors

- 400 Bad Request
- 401 Unauthorized
- 403 Forbidden
- 404 Not Found

5xx - Server errors

- 500 Internal Server Error

https://http.cat

In [15]:
#get status code for a "broken" link

r = requests.get('https://api.github.com/fake-ending')
r.status_code

403

Can also explictly ask for the returned format to be json as a method.

In [16]:
r.json()

{'message': "API rate limit exceeded for 65.207.79.74. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)",
 'documentation_url': 'https://developer.github.com/v3/#rate-limiting'}

#### A note on errors and exceptions with the Requests library

There are a number of exceptions and error codes you need to be familiar with when using the Requests library in Python.

- The Requests library will raise a ConnectionError exception if there is a network problem like a DNS failure, or refused connection.
- These are rare, but with invalid HTTP responses, Requests will also raise an HTTPError exception. 
- A Timeout exception will be raised if a request times out.
- If and when a request exceeds the preconfigured number of maximum redirections, then a TooManyRedirects exception will be raised

Now that we know a bit more about APIs and how to interact with them in python, let's actually investigate an API from scratch. 

To start, go over to the API documentation at: 

https://dev.socrata.com/foundry/data.cityofnewyork.us/fhrw-4uyv


<img src="311_api_docs.png">

## Make an initial API call to retrieve 311 complaints from a neighborhood or zip code of your choice in New York City.

### Using Requests

In [3]:
import requests
import pandas as pd

zip_code = '11004'

# can't figure out date ranges at the moment...
start_date = '2018-01-01T12:00:00'
end_date = '2018-02-01T12:00:00'

# create pull request based on parameters
# https://data.cityofnewyork.us/Social-Services/311-Service-Reque`sts-from-2010-to-Present/erm2-nwe9 click API

url = "https://data.cityofnewyork.us/resource/fhrw-4uyv.json?incident_zip={}".format(zip_code)

# do the pull
response = requests.get(url)
if response.status_code == 200:
    data = response.json()
else:
    print('Hit an error.')

## Briefly Explore the Structure of the Response You Received.

### Let's look at the output of our API call to NYC 311

In [20]:
print(type(data))

<class 'list'>


In [21]:
len(data)

1000

In [41]:
df_311 = pd.DataFrame.from_dict(data)
df_311

address_type agency                             agency_name         bbl  \
0    INTERSECTION    DOT            Department of Transportation         NaN   
1         ADDRESS    DPR      Department of Parks and Recreation  4087710037   
2         ADDRESS    DPR      Department of Parks and Recreation  4087340057   
3    INTERSECTION    DOT            Department of Transportation         NaN   
4         ADDRESS    DPR      Department of Parks and Recreation  4086840051   
5         ADDRESS   DSNY                Department of Sanitation  4087420125   
6    INTERSECTION    DOT            Department of Transportation         NaN   
7    INTERSECTION    DOT            Department of Transportation         NaN   
8         ADDRESS   DSNY                Department of Sanitation  4087400201   
9         ADDRESS    DEP  Department of Environmental Protection  4087400256   
10        ADDRESS    DEP  Department of Environmental Protection  4087400220   
11        ADDRESS    DEP  Department of Environmental Protection  4087410124   
12        ADDRESS    DEP  Department of Environmental Protection  4087230070   
13        ADDRESS    DEP  Department of Environmental Protection  4087410105   
14        ADDRESS    DEP  Department of Environmental Protection  4087410120   
15        ADDRESS    DEP  Department of Environmental Protection  4087390020   
16        ADDRESS   DSNY                Department of Sanitation  4087790033   
17   INTERSECTION    DOT            Department of Transportation         NaN   
18        ADDRESS    DEP  Department of Environmental Protection  4087240216   
19   INTERSECTION    DOT            Department of Transportation         NaN   
20   INTERSECTION    DOT            Department of Transportation         NaN   
21   INTERSECTION    DOT            Department of Transportation         NaN   
22   INTERSECTION    DOT            Department of Transportation         NaN   
23   INTERSECTION    DOT            Department of Transportation         NaN   
24   INTERSECTION    DOT            Department of Transportation         NaN   
25   INTERSECTION    DOT            Department of Transportation         NaN   
26   INTERSECTION    DOT            Department of Transportation         NaN   
27        ADDRESS    DOF                 Personal Exemption Unit  4087730019   
28   INTERSECTION    DOT            Department of Transportation         NaN   
29        ADDRESS    DOF                 Personal Exemption Unit  4084470001   
..            ...    ...                                     ...         ...   
970       ADDRESS    DOF                 Personal Exemption Unit  4086930026   
971       ADDRESS    DOF                    Property Exec Office  4087770033   
972       ADDRESS    DOF                    Property Exec Office  4087000020   
973       ADDRESS    DOF                 Personal Exemption Unit  4087100053   
974       ADDRESS    DOF                 Personal Exemption Unit  4086730012   
975       ADDRESS    DOF                 Personal Exemption Unit  4087430102   
976       ADDRESS    DOF                 Personal Exemption Unit  4087100053   
977       ADDRESS    DOF                 Personal Exemption Unit  4087150057   
978       ADDRESS    DOF               Commercial Exemption Unit  4086790073   
979       ADDRESS   DFTA                Department for the Aging  4087750043   
980       ADDRESS    DOF                 Personal Exemption Unit  4085350002   
981       ADDRESS    DOF                 Personal Exemption Unit  4085170002   
982       ADDRESS    DOF                     Correspondence Unit  4087000029   
983       ADDRESS    DOF                     Correspondence Unit  4087350053   
984       ADDRESS   DFTA                Department for the Aging  4086990059   
985       ADDRESS    DOF                      Condo or CoOp Unit  4085130020   
986       ADDRESS    DOF                    Property Exec Office  4087000048   
987       ADDRESS    DOF                     Correspondence Unit  4087560011   
988  

### Libraries for company-specific APIs

Google has its own libraries you can import to access their apis. <br>
NYC also uses the `sodapy` library, Socrata, to manage their open data. 


#### Extra credit for later if you want to play with it

In [ ]:
# Install before running 
!pip install sodapy

#### Tokenize yourself, B!

https://dev.socrata.com/foundry/data.cityofnewyork.us/fhrw-4uyv

Scroll down and click to sign up for an app token! This is instantaneous and simple. No credit cards required!


In [ ]:
token = '' 

import pandas as pd
from sodapy import Socrata

# Unauthenticated client only works with public data sets. Note 'None'
# in place of application token, and no username or password:
client = Socrata("data.cityofnewyork.us", token)

# Example authenticated client (needed for non-public datasets):
# client = Socrata(data.cityofnewyork.us,
#                  MyAppToken,
#                  userame="user@example.com",
#                  password="AFakePassword")

# First 2000 results, returned as JSON from API / converted to Python list of
# dictionaries by sodapy.
results = client.get("fhrw-4uyv", incident_zip = '11004', limit=1000)

In [13]:
type(results)

requests.models.Response

In [ ]:
len(results)

In [ ]:
results[0]

In [ ]:

df_1 = pd.DataFrame(results)

print(len(df_1))
print(df_1.columns)
df_1.head()

## Create a Pandas DataFrame of the Data From the Response

In [ ]:
df_2 = pd.DataFrame(data)

print(len(df_2))
print(df_2.columns)
df_2.head()

## Create a Histogram of the Complaint Types From Your Dataset

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
df_2.complaint_type.value_counts().plot(kind='barh', figsize=(8,12))

## If we have time:

https://github.com/toddmotto/public-apis


find a buddy, find a free api, get the key, and do a get. Try to transform the response into a df.

In [1]:
#   https://rickandmortyapi.com/
import requests
import pandas as pd
import numpy as np


In [2]:
url = "https://rickandmortyapi.com/api/character/"

# do the pull
response = requests.get(url)
if response.status_code == 200:
    data = response.json()
else:
    print('Hit an error.')

In [3]:
data['results'][0]['image']


'https://rickandmortyapi.com/api/character/avatar/1.jpeg'

In [4]:
df_rm_pictures = pd.DataFrame()
name_list = []
image_list = []

for i in range(25):
    
    url = "https://rickandmortyapi.com/api/character/?page={}".format(i+1)

    # do the pull
    response = requests.get(url)
    if response.status_code == 200:
        data = response.json()
    
        for rm_char in data['results']:
            name_list.append(rm_char['name'])
            image_list.append(rm_char['image'])
    
    
    else:
        print('Hit an error.')
    
    


df_rm_pictures['name'] = name_list
df_rm_pictures['image'] = image_list


In [40]:
type(data)

dict

In [5]:
df_rm_pictures.head()

name                                              image
0  Rick Sanchez  https://rickandmortyapi.com/api/character/avat...
1   Morty Smith  https://rickandmortyapi.com/api/character/avat...
2  Summer Smith  https://rickandmortyapi.com/api/character/avat...
3    Beth Smith  https://rickandmortyapi.com/api/character/avat...
4   Jerry Smith  https://rickandmortyapi.com/api/character/avat...

In [6]:
from IPython.display import display, Markdown, Latex

In [27]:
number_image = 1
i = np.random.choice(len(df_rm_pictures), number_image)


for rm_index in i:
    display(Markdown("{0}\n![should be {0}]({1})".format(df_rm_pictures.name[rm_index], df_rm_pictures.image[rm_index])))

Trandor
![should be Trandor](https://rickandmortyapi.com/api/character/avatar/363.jpeg)